src/auth/linking.ts

```diff
@@ -0,0 +1,299 @@
+  const rawSessionKey = await rsa.decrypt(encryptedSessionKey, ls.temporaryRsaPair.privateKey)
+  const sessionKey = await aes.importKey(utils.arrBufToBase64(rawSessionKey), { alg: SymmAlg.AES_GCM, length: 256 })
+
+  ls.sessionKey = sessionKey
+  ls.temporaryRsaPair = null
+
+  // Extract UCAN
+  const encodedUcan = await aes.decrypt(
+    json.msg,
+    ls.sessionKey,
+    {
+      alg: SymmAlg.AES_GCM,
+      iv: iv
+    }
+  )
+
+  const decodedUcan = ucan.decode(encodedUcan)
+
+  if (await ucan.isValid(decodedUcan) === false) {
+    throw new Error("Invalid closed UCAN")
+  }
+
+  if (decodedUcan.payload.ptc) {
+    throw new Error("Invalid closed UCAN: must not have any potency")
+  }
+
+  const sessionKeyFromFact = decodedUcan.payload.fct[0] && decodedUcan.payload.fct[0].sessionKey
+  if (!sessionKeyFromFact) {
+    throw new Error("Session key is missing from closed UCAN")
+  }
+
+  const sessionKeyWeAlreadyGot = utils.arrBufToBase64(rawSessionKey)
+  if (sessionKeyFromFact !== sessionKeyWeAlreadyGot) {
+    throw new Error("Closed UCAN session key does not match the one we already have")
+  }
+
+  return new Uint8Array(utils.randomBuf(6)).map(n => {
+    return n % 10
+  })
+}
+
+/**
+ * CONSUMER: NEGOTIATION (response)
+ * 
+ * Encrypt and publish the CONSUMER did and generated pin for verification by the PRODUCER.
+ * 
+ * @param pin 
+ * @returns 
+ */
+export const sendUserChallenge = async (pin: Uint8Array): Promise<void> => {
+  if (!ls.sessionKey) return
+  
+  const iv = utils.randomBuf(16)
+  const msg = await aes.encrypt(JSON.stringify({
+    did: await did.ucan(),
+    pin
+  }), ls.sessionKey, { iv, alg: SymmAlg.AES_GCM })
+
+  await publishOnChannel(
+    JSON.stringify({
+      iv: utils.arrBufToBase64(iv),
+      msg
+    })
+  )
+  nextStep()
+}
+
+/**
+ * PRODUCER: NEGOTIATION
+ * 
+ * PRODUCER receives the DID & challenge PIN from the CONSUMER.
+ * 
+ * @param data 
+ * @returns 
+ */
+export const handleUserChallenge = async (data: any): Promise<string | null> => {
+  if (!ls.sessionKey) return null
+
+  const { iv, msg } = JSON.parse(data)
+
+  console.debug("msg: " + msg)
+  if (!iv) {
+    throw new Error("I tried to decrypt some data (with AES) but the `iv` was missing from the message")
+  }
+
+  console.debug("decrypting msg")
+  return await aes.decrypt(msg, ls.sessionKey, {
+    alg: SymmAlg.AES_GCM,
+    iv
+  })
+  nextStep()
+}
+
+/**
+ * PRODUCER: DELEGATION
+ * 
+ * This step is user initiated, create the UCAN with delegate rights for the CONSUMER (and a WNFS read key).
+ * 
+ * @param data 
+ * @returns 
+ */
+export const sendReadKeyAndUcan = async (data: any): Promise<null> => {
+  return null
+}
+
+/**
+ * CONSUMER: DELEGATION
+ * 
+ * Receive the encrypted UCAN & readKey and store them.
+ * 
+ * @param readKey 
+ * @param ucan 
+ * @param username 
+ * @returns 
+ */
+export const handleLinkedDevice = async (readKey: string, ucan: string, username: string): Promise<null> => {
+  //
+  return null
```

Comment:
same as above.... not sure how best to surface this to the implementor. this data will actually come across the channel, so *not* via user action, but an implementor will probably want to have control over what happens to the readKey & UCan (and could potentially not even care about a readKey?)

I am leaning towards just having this be a DI impelemtation.